File: filesystem/v1_example.ts
import { createReadStream } from "node:fs";
import { resolve } from "node:path";
import Letta from "@letta-ai/letta-client";

const apiKey = process.env.LETTA_API_KEY;
if (!apiKey) throw new Error("LETTA_API_KEY is required");

const paperPath = resolve(process.env.PAPER_PDF ?? "memgpt.pdf");
const suffix = Date.now().toString(36);
const client = new Letta({
  apiKey,
  baseURL: "https://api.letta.com",
});

let agentId: string | undefined;
let folderId: string | undefined;
let attached = false;

const question = `Use the attached MemGPT paper to answer these questions. Cite short supporting excerpts from the paper for each answer.

1. What operating-system mechanism inspired MemGPT's virtual context management, and how does the paper map the LLM's main and external context to that mechanism?
2. Which two domains were used to evaluate MemGPT, and what new synthetic retrieval task did the paper introduce?`;

async function waitForUpload(fileId: string, targetFolderId: string) {
  const deadline = Date.now() + 5 * 60_000;

  while (Date.now() < deadline) {
    const file = await client.folders.files.retrieve(fileId, {
      folder_id: targetFolderId,
    });

    console.log(`upload status: ${file.processing_status ?? "unknown"}`);
    if (file.processing_status === "completed") return;
    if (file.processing_status === "error") {
      throw new Error(file.error_message ?? "File processing failed");
    }

    await new Promise((resolveDelay) => setTimeout(resolveDelay, 2_000));
  }

  throw new Error("Timed out waiting for file processing");
}

try {
  const folder = await client.folders.create({
    name: `memgpt-paper-v1-${suffix}`,
    description: "MemGPT paper used for filesystem document Q&A",
    embedding: "openai/text-embedding-3-small",
  });
  folderId = folder.id;

  const upload = await client.folders.files.upload(folder.id, {
    file: createReadStream(paperPath),
    name: "memgpt.pdf",
  });
  await waitForUpload(upload.id, folder.id);

  const agent = await client.agents.create({
    name: `memgpt-filesystem-example-${suffix}`,
    model: "openai/gpt-5.2",
    embedding: "openai/text-embedding-3-small",
    memory_blocks: [
      {
        label: "persona",
        value:
          "You are a careful research assistant. Answer document questions from attached files and quote the supporting text.",
      },
    ],
  });
  agentId = agent.id;

  // The current generated SDK takes folder ID first and agent ID in the body.
  await client.agents.folders.attach(folder.id, { agent_id: agent.id });
  attached = true;

  const response = await client.agents.messages.create(agent.id, {
    input: question,
  });

  for (const message of response.messages) {
    if (message.message_type === "assistant_message") {
      console.log(message.content);
    }
  }
} finally {
  if (process.env.KEEP_RESOURCES !== "1") {
    if (attached && agentId && folderId) {
      await client.agents.folders
        .detach(folderId, { agent_id: agentId })
        .catch(() => undefined);
    }
    if (agentId) await client.agents.delete(agentId).catch(() => undefined);
    if (folderId) await client.folders.delete(folderId).catch(() => undefined);
  }
}

File: filesystem/v1_example.py
import os
import time
from pathlib import Path

from letta_client import Letta

api_key = os.environ.get("LETTA_API_KEY")
if not api_key:
    raise RuntimeError("LETTA_API_KEY is required")

paper_path = Path(os.environ.get("PAPER_PDF", "memgpt.pdf")).resolve()
suffix = format(int(time.time() * 1000), "x")
client = Letta(api_key=api_key, base_url="https://api.letta.com")

agent_id = None
folder_id = None
attached = False

question = """Use the attached MemGPT paper to answer these questions. Cite short supporting excerpts from the paper for each answer.

1. What operating-system mechanism inspired MemGPT's virtual context management, and how does the paper map the LLM's main and external context to that mechanism?
2. Which two domains were used to evaluate MemGPT, and what new synthetic retrieval task did the paper introduce?"""


def wait_for_upload(file_id: str, target_folder_id: str) -> None:
    deadline = time.monotonic() + 5 * 60

    while time.monotonic() < deadline:
        uploaded_file = client.folders.files.retrieve(
            file_id,
            folder_id=target_folder_id,
        )
        status = uploaded_file.processing_status or "unknown"
        print(f"upload status: {status}")

        if status == "completed":
            return
        if status == "error":
            raise RuntimeError(uploaded_file.error_message or "File processing failed")

        time.sleep(2)

    raise TimeoutError("Timed out waiting for file processing")


try:
    folder = client.folders.create(
        name=f"memgpt-paper-v1-python-{suffix}",
        description="MemGPT paper used for filesystem document Q&A",
        embedding="openai/text-embedding-3-small",
    )
    folder_id = folder.id

    upload = client.folders.files.upload(
        folder.id,
        file=paper_path,
        name="memgpt.pdf",
    )
    wait_for_upload(upload.id, folder.id)

    agent = client.agents.create(
        name=f"memgpt-filesystem-python-{suffix}",
        model="openai/gpt-5.2",
        embedding="openai/text-embedding-3-small",
        memory_blocks=[
            {
                "label": "persona",
                "value": (
                    "You are a careful research assistant. Answer document "
                    "questions from attached files and quote the supporting text."
                ),
            }
        ],
    )
    agent_id = agent.id

    client.agents.folders.attach(folder.id, agent_id=agent.id)
    attached = True

    response = client.agents.messages.create(
        agent_id=agent.id,
        input=question,
    )

    for message in response.messages:
        if message.message_type == "assistant_message":
            print(message.content)
finally:
    if os.environ.get("KEEP_RESOURCES") != "1":
        if attached and agent_id and folder_id:
            try:
                client.agents.folders.detach(folder_id, agent_id=agent_id)
            except Exception:
                pass
        if agent_id:
            try:
                client.agents.delete(agent_id)
            except Exception:
                pass
        if folder_id:
            try:
                client.folders.delete(folder_id)
            except Exception:
                pass
        client.close()
